### PR TITLE
Reject non-canonical BN256 nullifiers

### PR DESCRIPTION
**Pull request by 1sgtpepper**
## Summary

Reject BN256 nullifiers that are not canonical field elements before they are passed to the verifier or written into pool state.

This adds a pool-level canonicality check for input nullifiers. Values greater than or equal to the BN256 scalar modulus now fail with `NonCanonicalPublicInput`.

## Changes

- Add `Error::NonCanonicalPublicInput`.
- Add a small helper to reject `U256 >= bn256_modulus`.
- Validate all input nullifiers before spent checks.
- Add a regression test for non-canonical nullifiers.

## Validation

Ran:

```sh
VERIFIER_VK_JSON=$PWD/deployments/testnet/circuit_keys/policy_tx_2_2_vk.json cargo test -p pool --locked -j 1 -- --nocapture
```

Result:

```text
9 passed; 0 failed
```

Also ran:

```sh
cargo fmt --package pool
git diff --check -- contracts/pool/src/pool.rs contracts/pool/src/test.rs
```

### CODE REVIEW

File: contracts/pool/src/pool.rs [hunk 1]
```diff
@@ -552,7 +562,9 @@ impl PoolContract {
             return Err(Error::UnknownRoot);
         }
         // 2. Nullifier checks (prevent double-spending)
+        let bn256_modulus = bn256_modulus(env);
         for n in proof.input_nullifiers.iter() {
+            Self::validate_bn256_public_input(&n, &bn256_modulus)?;
```
Comment (cupicmarko) [inline contracts/pool/src/pool.rs]: This is good, but also every `U256` field should be checked before converting to `Bn254Fr`, something like:

```rust
Self::validate_bn256_public_input(&proof.root, &modulus)?;
// ...

for n in &proof.input_nullifiers {
    Self::validate_bn256_public_input(&n, &modulus)?;
}
```

File: contracts/pool/src/test.rs
Comment (cupicmarko) [inline contracts/pool/src/test.rs:1]: Tests for boundary values could be added too (`r`, `r-1`)

File: contracts/pool/src/pool.rs [hunk 2]
```diff
@@ -51,6 +51,8 @@ pub enum Error {
     NotInitialized = 11,
     /// Arithmetic overflow occurred
     Overflow = 12,
+    /// Public input is not canonical in the BN256 scalar field
```
Comment (cupicmarko) [inline contracts/pool/src/pool.rs]:
```suggestion
    /// Public input is not canonical in the BN254 scalar field
```

**Review by cupicmarko — COMMENTED** [review]
PR correctly fixes non-canonical BN254 field elements before they reach the verifier, but it should be extended to validate all public inputs, not just nullifiers.

**Comment by 1sgtpepper** [comment]
Thanks for the review.

I’ve addressed the comments and added the requested boundary coverage. Could you take another look when you have a chance?

File: contracts/pool/src/test.rs [hunk 3]
```diff
@@ -453,3 +453,157 @@ fn transact_rejects_bad_public_amount() {
...
+
+    let asp_membership_root = setup.asp_membership_client.get_root();
+    let asp_non_membership_root = setup.asp_non_membership_client.get_root();
+
+    let proof = Proof {
+        proof: mk_mock_groth16_proof(&env),
+        root,
+        input_nullifiers: {
+            let mut v: Vec<U256> = Vec::new(&env);
+            let canonical_nullifier = U256::from_u32(&env, 123456789);
```
Comment (maksimryndin) [inline contracts/pool/src/test.rs]: the test is named `transact_rejects_non_canonical_nullifier` but here the var name is `canonical_nullifier`, isn't it a contradiction?

File: contracts/pool/src/test.rs [hunk 4]
```diff
@@ -453,3 +453,157 @@ fn transact_rejects_bad_public_amount() {
 
     assert!(pool.try_transact(&proof, &ext, &sender).is_err());
 }
+
+#[test]
+fn transact_rejects_non_canonical_nullifier() {
+    let env = test_env();
+    let setup = setup_test_contracts(&env);
+    let max = U256::from_u32(&env, 1000);
```
Comment (maksimryndin) [inline contracts/pool/src/test.rs]: I would have a more semantical name for this variable (and in the tests below) otherwise I don't understand its meaning in the env.register call below :)

**Review by maksimryndin — COMMENTED** [review]
Hey @Kuhai9801 ! thanks a lot for the contribution! 👍  Some small nit picks

**Comment by 1sgtpepper** [comment]
Thanks @maksimryndin, I’ve addressed the naming comments.

**Review by maksimryndin — APPROVED** [review]
amazing 👍 thank you @Kuhai9801 !

**Comment by maksimryndin** [comment]
@Kuhai9801 sorry,
"Commits must have verified signatures."
please refer to https://github.com/NethermindEth/stellar-private-payments/blob/main/CONTRIBUTING.md#commit-signing
you can squash your 5 commits into a single signed one

**Comment by 1sgtpepper** [comment]
Thanks, I squashed the branch into a single verified signed commit: ea6c9d7f774c913f79610cdcb9b3eecd2e75aaf0.
